exonum-java-binding/core/rust/src/runtime/service_factory_adapter.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright 2019 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use exonum::{
+    blockchain::Service,
+    helpers::fabric::{Command, CommandExtension, Context, ServiceFactory},
+};
+use runtime::{cmd::Run, config::Config, java_service_runtime::JavaServiceRuntime};
+use std::sync::{Once, ONCE_INIT};
+
+static mut JAVA_SERVICE_RUNTIME: Option<JavaServiceRuntime> = None;
+static JAVA_SERVICE_RUNTIME_INIT: Once = ONCE_INIT;
+static FIRST_INSTANCE_CREATED: Once = ONCE_INIT;
+
+/// Adapts current single-service interface of Exonum to multiple EJB services until dynamic
+/// services are implemented. Tracks the `JavaServiceRuntime` instantiation and command extension
+/// and represents factory for every user service relying on current `ServiceFactory` interface.
+pub struct JavaServiceFactoryAdapter {
+    name: String,
+    artifact_path: String,
+    should_extend_run_command: bool,
+}
+
+impl JavaServiceFactoryAdapter {
+    /// Creates new instance with given service name and path to the artifact.
+    pub fn new(name: String, artifact_path: String) -> Self {
+        JavaServiceFactoryAdapter {
+            name,
+            artifact_path,
+            // Only the first created instance should always extend the `Run` command.
+            should_extend_run_command: is_first_instance_created(),
+        }
+    }
+
+    // Creates new runtime from provided config or returns the one created earlier. There can be
+    // only one `JavaServiceRuntime` instance at a time.
+    fn get_or_create_java_service_runtime(config: Config) -> JavaServiceRuntime {
+        // Initialize runtime if it wasn't created before.
+        JAVA_SERVICE_RUNTIME_INIT.call_once(|| {
+            let runtime = JavaServiceRuntime::new(config);
+            unsafe {
+                JAVA_SERVICE_RUNTIME = Some(runtime);
+            }
+        });
+
+        unsafe { JAVA_SERVICE_RUNTIME.clone() }
+            .expect("Trying to return runtime, but it's uninitialized")
+    }
+
+    // Extend command only when required.
+    fn provide_run_command_extension(&self) -> Option<Box<dyn CommandExtension>> {
```

Comment:
Could we place the methods after their usages so that they are not scatterred randomly?

Reply:
As it doesn't belong to the trait implementation it's better to inline it in the parent method.